hy/model_patterns.py

```diff
@@ -0,0 +1,76 @@
+# Copyright 2018 the authors.
+# This file is part of Hy, which is free software licensed under the Expat
+# license. See the LICENSE.
+
+"Parser combinators for pattern-matching Hy model trees."
+
+from hy.models import HyExpression, HySymbol, HyKeyword, HyString, HyList
+from funcparserlib.parser import (
+    some, skip, many, finished, a, Parser, NoParseError, State)
+from functools import reduce
+from itertools import repeat
+from operator import add
+from math import isinf
+
+FORM = some(lambda _: True)
+SYM = some(lambda x: isinstance(x, HySymbol))
+STR = some(lambda x: isinstance(x, HyString))
+
+def sym(wanted):
+    "Parse and skip the given symbol or keyword."
+    if wanted.startswith(":"):
+        return skip(a(HyKeyword(wanted[1:])))
+    return skip(some(lambda x: isinstance(x, HySymbol) and x == wanted))
+
+def whole(parsers):
+    """Parse the parsers in the given list one after another, then
+    expect the end of the input."""
+    if len(parsers) == 0:
+        return finished >> (lambda x: [])
+    if len(parsers) == 1:
+        return parsers[0] + finished >> (lambda x: x[:-1])
+    return reduce(add, parsers) + skip(finished)
+
+def _grouped(group_type, parsers): return (
+    some(lambda x: isinstance(x, group_type)) >>
+    (lambda x: group_type(whole(parsers).parse(x)).replace(x, recursive=False)))
+
+def sb(*parsers):
```

Comment:
``sb`` I couldn't parse before reading the definition down here. We don't want to use too long name, because typing it would get annoying, but is there better than ``sb`` that isn't too long?

Reply:
`square` or `squareb`? `brackets` is perhaps obvious, but people sometimes use that word to mean bracketing characters in general, including parentheses.

Reply:
out of those, I like ``brackets`` best. But this is nothing we need to solve if good name doesn't come easy. I just thought that since I really like the pattern matching rules we have now, we could have obvious names for them. Now new developer has to check the definition to understand what's going on (not that the code is long).

Reply:
All right, I'll do a `filter-branch` to change it to `brackets`.